Code/GraphMol/Fingerprints/Fingerprints.cpp

```diff
@@ -173,6 +173,182 @@ return res;
 
 }  // end of anonymous namespace
 
+
+namespace utils {
+
+void buildDefaultAtomInvariants(const ROMol &mol,
+  std::vector<boost::uint32_t>& lAtomInvariants){
+
+  lAtomInvariants.reserve(mol.getNumAtoms());
+  for(ROMol::ConstAtomIterator atomIt=mol.beginAtoms();
+      atomIt!=mol.endAtoms();
+      ++atomIt){
+    unsigned int aHash = ((*atomIt)->getAtomicNum()%128)<<1 | (*atomIt)->getIsAromatic();
+    lAtomInvariants.push_back(aHash);
+  }
+}
+
+void enumerateAllPaths(const ROMol &mol, INT_PATH_LIST_MAP& allPaths,
+    const std::vector<boost::uint32_t> *fromAtoms,
+    bool branchedPaths, bool useHs,
+    unsigned int minPath, unsigned int maxPath){
+
+  if(!fromAtoms){
+    if(branchedPaths){
+      allPaths = findAllSubgraphsOfLengthsMtoN(mol,minPath,maxPath,
+          useHs);
+    }
+    else {
+      allPaths = findAllPathsOfLengthsMtoN(mol,minPath,maxPath,
+          useHs);
+    }
+  } else {
+    BOOST_FOREACH(boost::uint32_t aidx,*fromAtoms){
+      INT_PATH_LIST_MAP tPaths;
+      if(branchedPaths){
+        tPaths = findAllSubgraphsOfLengthsMtoN(mol,minPath,maxPath,
+            useHs,aidx);
+      }
+      else {
+        tPaths = findAllPathsOfLengthsMtoN(mol,minPath,maxPath,
+            true,useHs,aidx);
+      }
+      for(INT_PATH_LIST_MAP::const_iterator tpit=tPaths.begin();
+          tpit!=tPaths.end();++tpit){
+#ifdef VERBOSE_FINGERPRINTING
+        std::cerr << "paths from " << aidx << " size: " << tpit->first
+                  << std::endl;
+        BOOST_FOREACH (PATH_TYPE path, tpit->second) {
+          std::cerr << " path: ";
+          std::copy(path.begin(), path.end(),
+                    std::ostream_iterator<int>(std::cerr, ", "));
+          std::cerr << std::endl;
+        }
+#endif
+
+        allPaths[tpit->first].insert(allPaths[tpit->first].begin(),
+            tpit->second.begin(),tpit->second.end());
+      }
+    }
+  }
+}
+
+void identifyQueryBonds(const ROMol &mol,
+    std::vector<const Bond *>& bondCache,
+    std::vector<short>& isQueryBond){
+
+  bondCache.resize(mol.getNumBonds());
+  ROMol::EDGE_ITER firstB,lastB;
+  boost::tie(firstB,lastB) = mol.getEdges();
+  while(firstB!=lastB){
+    const Bond *bond = mol[*firstB].get();
+    isQueryBond[bond->getIdx()] = 0x0;
+    bondCache[bond->getIdx()]=bond;
+    if(isComplexQuery(bond)){
+      isQueryBond[bond->getIdx()] = 0x1;
+    }
+    if(isComplexQuery(bond->getBeginAtom())){
+      isQueryBond[bond->getIdx()] |= 0x2;
+    }
+    if(isComplexQuery(bond->getEndAtom())){
+      isQueryBond[bond->getIdx()] |= 0x4;
+    }
+    ++firstB;
+  }
+}
+
+std::vector<unsigned int> generateBondHashes(const ROMol &mol, boost::dynamic_bitset<>& atomsInPath,
+    const std::vector<const Bond *>& bondCache,
+    const std::vector<short>& isQueryBond,
+    const PATH_TYPE &path, bool useBondOrder,
+    std::vector<boost::uint32_t> *atomInvariants){
```

Comment:
looks like you need atomInvariants to be there. Please add an `PRECONDITION` for it.